packages/paid-storage-contract/src/contract.ts

```diff
@@ -0,0 +1,144 @@
+import {
+  Option,
+  StorageBalance,
+  StorageBalanceBounds,
+  StorageManagement,
+} from "near-contract-standards/lib";
+import {
+  AccountId,
+  Balance,
+  LookupMap,
+  NearBindgen,
+  NearPromise,
+  ONE_NEAR,
+  assert,
+  call,
+  near,
+  view,
+} from "near-sdk-js";
+
+const ACCOUNTS_MAP_PREFIX = "chain-deployed-ui";
+
+@NearBindgen({})
+class UserStorage implements StorageManagement {
+  accounts: LookupMap<StorageBalance>;
+
+  constructor() {
+    this.accounts = new LookupMap(ACCOUNTS_MAP_PREFIX);
+  }
+
+  @call({ payableFunction: true })
+  storage_deposit({
+    account_id,
+    registration_only,
+  }: {
+    account_id: AccountId;
+    registration_only: boolean;
+  }): StorageBalance {
+    const amount: Balance = near.attachedDeposit();
+    account_id = account_id ?? near.predecessorAccountId();
+
+    assert(amount > 0, "Deposit amount must be greater than zero");
+
+    // Initialize balance if account does not exist
+    if (!this.accounts.containsKey(account_id)) {
+      this.accounts.set(account_id, { total: 0n, available: 0n });
+    }
+
+    // Calculate new total balance
+    const currentValues = this.accounts.get(account_id);
+    this.accounts.set(account_id, {
+      ...currentValues,
+      total: currentValues.total + amount,
+    });
+
+    if (!registration_only) {
+      // Increase available balance if not just registration
+      const currentValues = this.accounts.get(account_id);
+      this.accounts.set(account_id, {
+        ...currentValues,
+        available: currentValues.available + amount,
+      });
+    }
+
+    return this.accounts.get(account_id);
+  }
+
+  @call({})
+  storage_withdraw({ amount }: { amount?: bigint }): StorageBalance {
+    if (!amount) {
+      throw Error("The amount should be greater than zero");
+    }
+
+    amount = BigInt(amount);
```

Comment:
Is this required if `amount` is already typed as `bigint`? Does it still come in as a string?

Reply:
yeah, I've had an issue that before adding this line the contract was saying that it's a wrong type (I guess it was string instead). So I've found an inspiration here -> https://github.com/near/near-sdk-js/blob/develop/packages/near-contract-standards/src/fungible_token/core_impl.ts#L314 and adding this casing has resolved the issue.